File: ALL/Algo_Edit_Distance.cpp
#include <bits/stdc++.h>
using namespace std;

int min(int x, int y, int z) { 
    return min(min(x, y), z); 
}

int editDist(string str1, string str2) {
    int m = str1.length();
    int n = str2.length();
    
    vector<vector<int>> dp(m + 1, vector<int>(n + 1, 0));

    for (int i = 0; i <= m; i++) {
        for (int j = 0; j <= n; j++) {
            if (i == 0) {
                dp[i][j] = j;
            } else if (j == 0) {
                dp[i][j] = i;
            } else if (str1[i - 1] == str2[j - 1]) {
                dp[i][j] = dp[i - 1][j - 1];
            } else {
                dp[i][j] = 1 + min(dp[i][j - 1], dp[i - 1][j], dp[i - 1][j - 1]);
            }
        }
    }
    
    return dp[m][n];
}

int main() {   
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    string str1;
    string str2;
    
    cin >> str1 >> str2;

    cout << editDist(str1, str2);

    return 0;
}

File: ALL/Algo_Edit_Distance.py
def levenshtein_distance(str1, str2):
    cost = 1
    m = len(str1)
    n = len(str2)
    
    dp = [[0] * (n + 1) for _ in range(m + 1)]

    for i in range(m + 1):
        dp[i][0] = i * cost 

    for j in range(n + 1):
        dp[0][j] = j * cost 

    for i in range(1, m + 1):
        for j in range(1, n + 1):
            insert_cost = dp[i][j - 1] + cost
            delete_cost = dp[i - 1][j] + cost 
            replace_cost = dp[i - 1][j - 1] + (cost if str1[i - 1] != str2[j - 1] else 0)

            dp[i][j] = min(insert_cost, delete_cost, replace_cost)

    return dp[m][n]

str1 = input()
str2 = input()

result = levenshtein_distance(str1, str2)
print(result)
